### ISSUE TITLE
Modifying mandatory property doesn't work

### ISSUE THREAD

**Issue by eden-jh** [umbraco/Umbraco-CMS#12421]
### Which *exact* Umbraco version are you using? For example: 9.0.1 - don't just write v9

9.4.3

### Bug summary

Dynamically setting the `validation.mandatory` property of umbProperty's `property` object to true toggles the appearance of the label, but doesn't validate correctly, and breaks other client-side validation (for example, maxChars on a textbox) within the same block.

### Specifics

Changing the value of `validation.mandatory` to true (when the server setting is false) for an umbProperty results in all the client-side validation being ignored when you save and publish (so you're able to publish content with client-side validation errors). This does not impact server-side errors (and in fact client-side errors work mostly as expected if there's also a server-side error). 

This is the result of the `valPropertyMsg` directive and the specific way it handles clearing errors. Basically it uses the `isMandatory` variable to determine whether or not to clear errors from a property that has changed:

1. `isMandatory` is set in the `link` function and is local to that function. It is never updated, and there's no way to modify it.
2. `valPropertyMsg` has a `checkValidationStatus` that is called whenever the validation status of a control changes (which includes changing the value of `mandatory`)
3. In `checkValidationStatus`, when the directive determines that an empty, non-mandatory property has errors, it calls `$setValidity()` (with no arguments) on all of its controls to clear the error state. This does clear the control's `$error`, but it also adds an error key of `undefined` to its `$pending` property.
4. When `$pending` is truthy, $setValidity changes the form control's `$invalid` and `$valid` to undefined, which is propagated up to the parent form, even when other properties in the same form have `$invalid` as true
5. In the `submitForm` function of `formHelper`, publishing is only stopped if the form's `$invalid` property is truthy

https://github.com/umbraco/Umbraco-CMS/blob/78786e942fa302b011ed9e4782a28d1f32f6f585/src/Umbraco.Web.UI.Client/src/common/directives/validation/valpropertymsg.directive.js#L219-L232

### Steps to reproduce

1. Create a new document type with 2 textbox properties (I'll call them A and B), each with their own textbox configuration. Make sure neither property is mandatory. Leave Maximum allowed characters blank for A, and set it to 20 for B.
2. Save your changes
3. In the content section, add a content node of your new document type
4. In the console, get a jQuery/jQLite object for the `umb-property` element for Property A (I do this by right-clicking on the element in Inspector, then clicking Use in Console (for Firefox)/Store as as global variable (for Chrome) then wrapping that variable in`$()`)
5. Using this object, enter `obj.scope().property.validation.mandatory = true` in the console
6. Click anywhere on the Umbraco page (just to move focus out of dev tools) and notice Property A now shows an asterisk indicating that it's mandatory
7. In your console, use the property object to add a listener for formSubmitting so it will check whether the property is empty and set validity as appropriate:
```
obj.scope().$on("formSubmitting",function(){

   var validity = (!obj.scope().property.value || obj.scope().property.value == "");
   var controls = obj.find('ng-form').first().scope().propertyForm.$getControls(); 
   for(var i = 0; i < controls.length; i++) {
	var control = controls[i];
	control.$setValidity('required', validity);
   }
})
```
8. In property B, enter more than 20 characters. A message below the field will tell you you've exceeded the character limit
9. Hit the Save and publish button

### Expected result / actual result

Expected result: Validation errors on property A (for being an empty mandatory field) and property B (for exceeding the maximum number of characters) should appear and should prevent the content from being published.

There's currently no other way of dynamically changing a property to be mandatory, so if that makes more sense as a solution then that's fine too (as long as it can be applied externally to any property).

Actual result: Property A very fleetingly shows a "Required" message (with no error styles present) before that disappears.   Property B continues to show the message that the character limit was exceeded. Neither property has the classes used for error styles, and the content publishes successfully.

**Comment by eden-jh** [comment]
I'm also happy to submit a feature request for an external way of modifying `validation.mandatory` if that would be more appropriate.

**Comment by nul800sebastiaan** [comment]
I am just trying to understand what you're actually trying to achieve here @eden-jh - I am assuming your editors are not typing things in the browser console to change the validation. So what is it you're really trying to do? And how(-ish)?

**Comment by eden-jh** [comment]
Right, editors won't be typing in the browser console; I just used that as an example because I though it would be simplest for the purposes of reproduction. 

The validation is being changed by a Conditional Logic property editor, which can be configured to show, hide, or show and require another property in the same element/content type based on the value of other properties in the same content/element type. If a condition is met and the property should be hidden, or if it's not met and the property should be shown, it's hidden with display: none. For example, there could be a condition to show and require an Image property if a Media Type property value is equal to Image.

The hiding and showing works fine, but I need to be able to make a property mandatory when it's showing. I can't do that using the property settings of the content type, because then the server validation will find an error when that mandatory property is hidden. So that's why the validation has to be changed in the JS.

**Comment by nathanwoulfe** [comment]
@eden-jh I'm not sure I'd classify this as a bug, it's seems like intentional behaviour to avoid ending up with mismatched in validation between the client and the server. Making a server-side optional property mandatory on the client wouldn't cause issues when saving/publishing, but the inverse certainly would, so I'm guessing this was intentional to prevent giving devs too much rope.

I could also be entirely wrong, and this could be an oversight or genuine bug...

Personally, I'd approach this with a custom property editor, in which I could manage all validation logic, rather than trying to modify validation on other property editors (which may run into scope access issues when debug is disabled).

**Comment by eden-jh** [comment]
@nathanwoulfe I see what you mean. I wasn't entirely sure it counted as a bug, but I'm happy to create a separate feature request issue for this if that would be more appropriate. 

I considered custom property editors for the fields that are conditionally mandatory, but I don't think that's a good fit here for a few reasons (mainly that I'd like to avoid having to create custom property editors that would otherwise be functionally identical to existing ones, and that core directives use `validation.mandatory` to determine HTML so the only alternative to modifying it in JS anyway would be to recreate a bunch of them).

Based on the comment where `isMandatory` is declared, it doesn't appear to have been a deliberate choice to prevent changing the value, just a way of dealing with an undefined `validation` property:
https://github.com/umbraco/Umbraco-CMS/blob/78786e942fa302b011ed9e4782a28d1f32f6f585/src/Umbraco.Web.UI.Client/src/common/directives/validation/valpropertymsg.directive.js#L46-L47

Would you be open to an edit that handles the undefined value while still checking `currentProperty.validation.mandatory` if it exists?

**Comment by nul800sebastiaan** [comment]
Thanks @nathanwoulfe for the feedback here, this is indeed not an intended behavior. 

> it doesn't appear to have been a deliberate choice to prevent changing the value,

That is probably correct, this part of the code is now over 6 or 7 years old and we've learned a LOT about how people use and abuse the system. 

The feature you're seemingly looking for is not something that we'll be working on at the moment, but it would be good input for the new backoffice UI team. 

I have sent this off to them. If the team has any more questions they'll be asking them here.

Since we don't plan to work in this for the current backoffice I will close this issue, but it's still on the list of ideas for future implementations, that is what the `status/idea` label is for so we can easily find all the input again in the future. When we do decide to work on this we can reopen this issue again.

Thanks for the input so far!

**Comment by nul800sebastiaan** [comment]
Forgot to link to the new backoffice team blog post: https://umbraco.com/blog/introducing-the-new-new-backoffice-community-team/
